__tests__/app.test.js

```diff
@@ -0,0 +1,254 @@
+        .expect(200)
+        .then((response) => {
+          const comments = response.body.comments;
+          expect(comments).toBeSorted({ key: "created_at", descending: true });
+        });
+    });
+    test("Returns 400 Bad Request if given an article_id that is invalid", () => {
+      return request(app)
+        .get("/api/articles/potato/comments")
+        .expect(400)
+        .then((response) => {
+          expect(response.body.msg).toBe("Bad Request");
+        });
+    });
+    test("Returns 404 not found if given a valid article_id that does not exist", () => {
+      return request(app)
+        .get("/api/articles/9999/comments")
+        .expect(404)
+        .then((response) => {
+          expect(response.body.msg).toBe("Not Found");
+        });
+    });
+    test("Returns an empty array if given a valid, existing article_id which has no comments", () => {
+      return request(app)
+        .get("/api/articles/2/comments")
+        .expect(200)
+        .then((response) => {
+          expect(response.body.comments).toEqual([]);
+        });
+    });
+  });
+  describe("POST", () => {
+    test("Can post an article at valid id by sending body and username", () => {
+      const postData = { body: "example body", username: "butter_bridge" };
+      return request(app)
+        .post("/api/articles/2/comments")
+        .send(postData)
+        .expect(201)
+        .then((response) => {
+          const comment = response.body.comment;
+          expect(typeof comment.comment_id).toBe("number");
+          expect(typeof comment.article_id).toBe("number");
+          expect(typeof comment.votes).toBe("number");
+          expect(typeof comment.body).toBe("string");
+          expect(typeof comment.author).toBe("string");
+          expect(typeof comment.created_at).toBe("string");
+        });
+    });
+    test("Returns 400 Bad Request if given an article_id that is invalid", () => {
+      const postData = { body: "example body", username: "butter_bridge" }
+      return request(app)
+      .get("/api/articles/soap/comments")
+      .send(postData)
+        .expect(400)
+        .then((response) => {
+          expect(response.body.msg).toBe("Bad Request");
+        });
+    });
+    test("Returns 404 not found if given a valid article_id that does not exist", () => {
+      const postData = { body: "example body", username: "butter_bridge" }
+      return request(app)
+        .post("/api/articles/12345/comments")
+        .send(postData)
+        .expect(404)
+        .then((response) => {
+          expect(response.body.msg).toBe("Not Found");
+        });
+    });
+    test("Returns a 401 when given an invalid username", ()=>{
+      const postData = { body: "example body", username: "not_a_user" }
+      return request(app)
+      .post("/api/articles/2/comments")
+        .send(postData)
+        .expect(401)
+        .then((response) =>{
+          expect(response.body.msg).toBe('Unauthorised User')
```

Comment:
I don't quite agree with the choice of status code and error message here -
A 401 implies that the user does exist but doesn't have permission - but this user doesn't even exist at all! I'd recommend a 404 for this reason.